chat/utils.py

```diff
@@ -14,13 +18,22 @@ def run_llm(chat_history, context):
     models = ["gemini-1.5-flash-latest", 'gemini-1.5-pro']
     model = genai.GenerativeModel(models[0], system_instruction=context)
 
-    # TODO: need to check if the chat history is too long and make it shorter
-    messages = [*chat_history]
+    messages = [
+        {"role": msg["role"], "parts": msg["parts"], "pk": msg.get("pk"), "no_delete": msg.get("no_delete", False)}
+        for msg in chat_history
+    ]
     token_count = model.count_tokens(messages)
 
-    # check the number of tokens
-    # if too big, new message = [history (minus first message)] (keep checking until under a milly)
-    # in the future we want to mark some messages as no delete
+    # TODO: in the future we want to mark some messages as no delete
+    permanent_messages = [msg for msg in messages if msg.get('no_delete', False)]
+    messages_to_remove = [msg for msg in messages if
+                          not msg.get('no_delete', False) and not Message.objects.filter(pk=msg['pk']).exists()]
+
+    while token_count.total_tokens > MAX_TOKENS:
+        messages_to_remove = messages_to_remove[1:]  # Remove the oldest message from the history
+        token_count = model.count_tokens(permanent_messages + messages_to_remove)
+
+    messages = permanent_messages + messages_to_remove
```

Comment:
This is going to break the order of the messages. 
We also dont need to count the tokens on line 25 because we already have that data.
We only need to check the size of the prompt here, the return the result, prompt_token_size and response_token_size.

Then you can go with a while loop but only need to check total size of all all parts added together.
Then pop the first element in the list until it is under size.
If you run into a no delete message, pop the second message (keep going until you find a deleteable message)